src/nodes.coffee

```diff
@@ -762,8 +748,43 @@ exports.Block = class Block extends Base
     return nodes[0] if nodes.length is 1 and nodes[0] instanceof Block
     new Block nodes
 
-  astProperties: ->
-    expressions: @expressions.map (child) => child.ast()
+  asExpressionStatementAst: (ast) ->
+    # return ast if ast.type in ['ImportDeclaration']
+
+    Object.assign
+      type: 'ExpressionStatement'
+      expression: ast
+    ,
+      extractAstLocationData ast
+
+  getExpressionAst: (expression) ->
+    ast = expression.ast()
+    # return ast if expression.isStatement o
+    @asExpressionStatementAst ast
+
+  bodyToAst: ->
+    @getExpressionAst(expression) for expression in @expressions
+
+  rootToAst: ->
+    programLocationData = @astLocationData()
+
+    programAst = Object.assign
```

Comment:
Rather than a `Block` having two “modes,” root and body, should we perhaps split this class up into a base and children? Like `BlockBase` as a parent for `Root` and `Block`?

Reply:
I also wonder if `file.program` should be handled outside of the node classes, since that’s more like part of the file itself than part of the AST? We could return a root with `expressions` like we were before, or perhaps rename `expressions` to `body` in the AST, and then the parent `file.program` could be added outside of this like in `coffeescript.coffee` or somewhere else.

Reply:
> Rather than a `Block` having two “modes,” root and body, should we perhaps split this class up into a base and children? Like `BlockBase` as a parent for `Root` and `Block`?

Ok introduced a new `Root` class, but made it a _parent_ of the root `Block` rather than a subclass variation of it

> I also wonder if `file.program` should be handled outside of the node classes, since that’s more like part of the file itself than part of the AST?

Handling the root `File` AST node outside of the node classes makes a certain amount of sense conceptually but in practice it seems like it should stay in the node classes since (a) that `File` node needs to have `comments` attached to it and (b) it seems weird not to be able to use our AST methods to generate that root AST node

So I think this makes the most sense - now we can actually use the standard AST methods for both the `File` AST node (which corresponds to the `Root` node class) and the `Program` AST node (which corresponds to the root block, ie the `body` property of the `Root` node class instance), so from an AST-generation perspective it's a significant cleanup

This also cleans up the root case when compiling JS a bit - now `Root::compileNode()` does most of what `Block::compileRoot()` did previously, and `Block::compileToFragments()` is gone (since `Root::compileNode()` can directly call the still-special `Block::compileRoot()` on its `@body` since it already knows that it's the root block)

The interesting thing I ran into was that producing/expecting this new `Root` node was breaking the existing interfaces of `CoffeeScript.nodes()` and the corresponding call to `.compile()` (on the root node returned by `CoffeeScript.nodes()`)

For `.compile()`, I added a backwards-compatibility override `Block::compile()` which checks if it needs to wrap itself in a `new Root`. So that way any existing code that expects to be able to call `.compile()` on a "root" `Block` instance will still work

For `CoffeeScript.nodes()` it was slightly more interesting. We could just return the new `Root` instance (as the root of the node class instance tree), and then clean up internal references (eg in `repl.coffee`, a few tests) that expect the return value of `CoffeeScript.nodes()` to be a `Block` instance

But then any third-party consumers of the `CoffeeScript.nodes()` API that expect it to return a `Block` would break. So what I did was return the "root" `Block` instance (ie the `body` of the `Root` instance) from `CoffeeScript.nodes()`. This is a little weird in that it makes the `Root` instance impossible to obtain a reference to via `CoffeeScript.nodes()`, but I think it's preferable to introducing breakage for consumers of `CoffeeScript.nodes()`?